Source/WebCore/rendering/RenderTextControlSingleLine.cpp

```diff
@@ -198,7 +199,9 @@ void RenderTextControlSingleLine::layout()
         auto* containerBox = innerTextRenderer ? innerTextRenderer : innerBlockRenderer ? innerBlockRenderer : containerRenderer;
         if (containerBox) {
             // Center vertical align the placeholder content.
-            auto logicalTop = placeholderTopLeft.y() + (containerBox->logicalHeight() / 2 - placeholderBox->logicalHeight() / 2);
+            auto placeholderLineBox = InlineIterator::firstLineBoxFor(downcast<RenderBlockFlow>(*placeholderBox));
```

Comment:
Are you sure getting the height of the first line box is always enough?
What if placeholder had a mixture of English & Japanese/Chinese?
Do we still use the correct height?

Reply:
I tested with a mixture of English/Chinese characters, and it appears to layout correctly.

Is there a specific behavior you’re concerned about?

(cc @alanbujtas)

Reply:
Okay. Have you tried Hebrew / Arabic & English? I just wanna make sure that just using the first box doesn't result in misalignment.

Reply:
Very nice!

Reply:
what kind of misalignment are you thinking of? The line box by definition contains all the inline content for that particular line regardless of their attributes etc. The only way this may fail if  the "placeholder" text wrapped into multiple lines and for that the content height would be: (lastLineBox.contentLogicalBottom() - firsLineBox.contentLogicalTop()). As for the actual height value, this patch specifically calls into the _content_logicalHeight() API which is the enclosing geometry for the line (and not the line box's height value that we use to size the containing block container). The commit message should be a bit more clear about that actually.
"use the placeholder's first line box height "
This is not the line box height, this is the height of the content in the first line box.

Reply:
Line boxes can be split into multiple boxes with bidirectional text.

Reply:
how do you split a line box with bidirectional text? are you referring to this line box https://www.w3.org/TR/css-inline-3/#line-box or something else (maybe inline box)?

Reply:
The "first box" refers to a _line_ box, not the content inside the line e.g. an inline box.

Reply:
Oh I see. Sorry, I got confused between line boxes & inline boxes.